STN-568: Views Dark Color Inversion

rkelbel opened the pull request:
# READY FOR REVIEW

## Summary
Use an existing pairing to reverse color on specific elements in Colorful and Traditional themes.
This PR should apply dark mode variants for views, using both the theme setting for dark inversion and the `hb-dark-pattern` utility class directly applied to the view.

JIRA Card: https://sparkbox.atlassian.net/browse/STN-568
Figma Link: [(Upcoming Events & Seminars Section)](https://www.figma.com/file/toNZnFLadtpNpG5dPzNGWz/%E2%9A%99%EF%B8%8F-Colorful-Theme---Working?node-id=248%3A0)

## Need Review By (Date)
12/16/20

## Urgency
low

**For this PR you will need to test both the Theme Dark Inversion setting (which can be found at Appearance > [Theme] > Settings) and the `hb-dark-pattern` utility class for any views you are testing on.**

## Steps to Test
1. `npm run test`
2. To test the dark mode color inversion toggle for views.
- Switch on the dark inversion setting in the theme appearance settings.
- From here you will need to test a few different view types, such as: events, news, people, publications, courses and research, etc.
- Some pages you can use to view these views listed below:
http://sparkbox-sandbox.suhumsci.loc/qa-views (`hb-pill`s seems to be broken in light and dark mode, not from work done on this card.)
_Courses:_
http://sparkbox-sandbox.suhumsci.loc/courses
http://sparkbox-sandbox.suhumsci.loc/courses-0
_Events:_
http://sparkbox-sandbox.suhumsci.loc/events
http://sparkbox-sandbox.suhumsci.loc/events/past-events
_People:_
http://sparkbox-sandbox.suhumsci.loc/people/people-flexible-image-classes
http://sparkbox-sandbox.suhumsci.loc/people/people-grid
_News:_
http://sparkbox-sandbox.suhumsci.loc/example-content/news
http://sparkbox-sandbox.suhumsci.loc/news
_Publications:_
http://sparkbox-sandbox.suhumsci.loc/publications/book-shelf
http://sparkbox-sandbox.suhumsci.loc/publications/publication-table
_Seminars:_
http://sparkbox-sandbox.suhumsci.loc/publications/book-shelf

- You can also create your own page and add Views blocks in manually through Paragraphs, if you'd like to test that way as well.
3. Once you are done testing different views with the toggle setting on the first theme type, you'll want to swap it over to Colorful or Traditional now and test those pages again depending on which theme you tested first.
4. To test `hb-dark-pattern` utility class in views individually. (make sure the toggle for the theme is turned off)
- To add the utility class you'll need to find the first view you'd like to test and click the "Edit View" contextual link for it.
- From there add the `hb-dark-pattern` class to the view by going to the "CSS class:" setting under the Views "Other:" section.
- Once you've added the class, click Apply on the setting window, then Save the view.
- The view should now look the same as what it looked like while you were testing the Dark Mode theme setting in the appearance settings.
- Repeat these steps for any additional views you'd like to test using the list above or the flexible page you may have created.
- Check the views in both Colorful and Traditional themes and color pairings.

**Note: The site search view was not tested or styled as it is specifically called out in the card as not included in this work.**

## PR Checklist
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)
- [Sparkbox PR Checklist](../docs/SparkboxPRChecklist.md)

Commit:
feat(STN-568): dark color inversion for views

heatherdesigns reviewed (commented):
Looking good!!

![source](https://user-images.githubusercontent.com/12678977/102242640-43fd7f00-3ec8-11eb-9302-70835fc8935d.gif)

I noticed a couple things we should address for consistency. The views should look the same when the `hb-dark-pattern` or the dark inversion toggle is applied. Here's a screenshot of a view on the homepage:

![home page](https://user-images.githubusercontent.com/12678977/102242872-8aeb7480-3ec8-11eb-97ed-57c3d99202ca.png)

When `hb-dark-pattern` is applied, the arrow should be white.
I noticed the pipe beside the categories is white or light (I can't tell which. 😂) when `hb-dark-pattern` is applied and is secondary green with the toggle. Let's make them the same. I really like the lighter version shown on `hb-dark-pattern`.

<hr>

The text in the pill that is in this view on the homepage is inaccessible. I looked at pills on the QA Views page and they don't look like this. This view is inside a well so I'm wondering if that may be causing the issue.

![homepage pill with utility](https://user-images.githubusercontent.com/12678977/102243456-31377a00-3ec9-11eb-88bb-4798092fa8ab.png)

<hr>

heatherdesigns reviewed (commented):
A couple more things:

**Tables.** This one is interesting! The colors of the rows are different when `hb-dark-pattern` and the dark color inversion toggle are applied. This is the Courses page.

![table](https://user-images.githubusercontent.com/12678977/102249401-5380c600-3ed0-11eb-9bc0-3fd5b7b61f0d.png)

Here's a screenshot of Courses without `hb-dark-pattern` or the dark inversion. When looking at this, the styles on the dark color inversion shown above are correct.

![courses before](https://user-images.githubusercontent.com/12678977/102249453-6398a580-3ed0-11eb-8cd4-e226fe98cacd.png)

<hr>

**Events.** I noticed that when a category tag does not have a link, it has light text. All category tags (link or text) should be the same color. Let's go with white as seen on the `Conference` category in this screenshot. Categories with links add an underline on hover, so I think we'll be ok.

I'm wondering if the icons should also be white. 🤔 But it isn't an accessibility issue so we can always change them in a refinement.

![event - category link versus no link](https://user-images.githubusercontent.com/12678977/102249934-e91c5580-3ed0-11eb-8eff-8759730fcdaf.png)

<hr>

**Pager.** The next and last links when `hb-dark-pattern` is applied should be the same color as the numbers. Toggle is correct.

![pager - more and next links](https://user-images.githubusercontent.com/12678977/102250619-c2125380-3ed1-11eb-8365-95177629d48e.png)

heatherdesigns reviewed (commented):
A couple more things. The feedback for these items includes both the `hb-dark-pattern` utility class and dark inversion toggle. These items look the same on both.

**News** - horizontal view. This one was super hard to notice. I almost missed it! 🙈 
The border in between the date and the website link should be white.

![news - border](https://user-images.githubusercontent.com/12678977/102263724-b24f3b00-3ee2-11eb-8192-0fbe8a4e5f38.png)

<hr>

Heading links in all views. The light bottom border color on heading links contrasting with the heading text color is kinda driving me crazy. 😆 I know it was there when you demo'ed in stand up earlier this week and Tori didn't request for it to be changed, but I wanted to see what it would look like if the border-bottom and text where the same color. I think it's easier to read and looks better. What do you think? You can always blame the change on me. 🤣 I added `border-bottom-color: initial` to make the border-bottom color match the text.

![view heading links](https://user-images.githubusercontent.com/12678977/102264289-6bae1080-3ee3-11eb-8b09-1d0e65108f2f.png)

<hr>

Vertical date stacked cards with `hb-raised-cards` inside of a well need some love. This is on http://sparkbox-sandbox.suhumsci.loc/texts-cards-row-wells

This is on Colorful:
![vertical stacked card in well](https://user-images.githubusercontent.com/12678977/102264777-16beca00-3ee4-11eb-961b-fbd9d40108ca.png)

And this is Traditional: (that drop shadow :eyes:)
![Screen Shot 2020-12-15 at 2 45 21 PM](https://user-images.githubusercontent.com/12678977/102264843-30601180-3ee4-11eb-816c-6320ce0b060d.png)

rkelbel commented:
> **Tables.** This one is interesting! The colors of the rows are different when `hb-dark-pattern` and the dark color inversion toggle are applied. This is the Courses page.
> 
> Here's a screenshot of Courses without `hb-dark-pattern` or the dark inversion. When looking at this, the styles on the dark color inversion shown above are correct.

I updated the dark pattern and inversions for the tables, adjusted the row colors to the original color. I also made the inversion colors match for consistency.  There was previous work done that altered the header color, so I added inversion to that rule to keep the colors consistent.

rkelbel commented:
> I'm wondering if the icons should also be white. 🤔 But it isn't an accessibility issue so we can always change them in a refinement.

- I swapped the icons to white, let me know if we want to swap them back.

rkelbel commented:
> **News** - horizontal view. This one was super hard to notice. I almost missed it! 🙈
> The border in between the date and the website link should be white.

I did miss this, good catch!

rkelbel commented:
> Heading links in all views. The light bottom border color on heading links contrasting with the heading text color is kinda driving me crazy. 😆 I know it was there when you demo'ed in stand up earlier this week and Tori didn't request for it to be changed, but I wanted to see what it would look like if the border-bottom and text where the same color. I think it's easier to read and looks better. What do you think? You can always blame the change on me. 🤣 I added `border-bottom-color: initial` to make the border-bottom color match the text.

I agree with this, I've updated the border bottom to match the text.

rkelbel commented:
> Vertical date stacked cards with `hb-raised-cards` inside of a well need some love. This is on http://sparkbox-sandbox.suhumsci.loc/texts-cards-row-wells

hb-raised-cards were interesting since by default raised card gets a white background, but the background behind it inverted.  I probably took the difficult route and used the white background upon the inverted backdrop instead of inverting the whole raised card. I think I have all the issues squared away. I'll be interested in feedback on this use case, as it likely should be used very sparingly for raised cards.

heatherdesigns reviewed (commented):
## Round 2 🎉 
Feedback for `hb-dark-pattern` and the dark inversion toggle

*Table header*
~~- [ ] The table header should keep the same color as when there is no inversion toggle or utility class applied. The screenshot below is of the Colorful theme with Ocean. In this example, both table headers should be the teal blue. I only see this issue in Colorful. Traditional looks ok.~~

### EDIT / UPDATE!
After reviewing your code, I doubled checked the color pairing spreadsheet. What you have is correct and reflects that which is in the color pairing spreadsheet. No changes need to be made to the table header. 

![table header](https://user-images.githubusercontent.com/12678977/102365165-407bfd80-3f85-11eb-9622-2aac1b998758.png)

<hr>

*Arrow icons in views on Colorful*
- [x] `hb-dark-pattern` arrows need to be updated.

![utility arrows on colorful](https://user-images.githubusercontent.com/12678977/102365334-6ef9d880-3f85-11eb-898c-cc1206cdc5a3.png)

<hr>

*Arrows on hover*
- [x] The weird bumpy transition of the arrows in views on hover is a result of the additional padding. I didn't investigate further into why it's happening. This is on both Colorful and Traditional.

<hr>

*Headings in views on Traditional*
- [x] When hovering over a heading in Traditional, the border-bottom should be the same color as the text. Sorry I forget to put this in yesterday's feedback. 🙈 

![heading hover](https://user-images.githubusercontent.com/12678977/102379088-812f4300-3f94-11eb-908d-4bb3d5f15de6.png)

<hr>

*Date in event card*
- [x] When the `hb-dark-pattern` utility class is applied to the views on the event pages, the date in the decorative calendar square (tile) is white and blends into the background. It looks like the styles causing this are pretty generic as it's coming from `.views-element-container .hb-dark-pattern .field-content`.

I left a comment in `card.scss` for a potential solution.

Let me know if you aren't seeing the same thing locally and we can pair. 🍐 

![traditional w utility class](https://user-images.githubusercontent.com/12678977/102382182-24358c00-3f98-11eb-908c-b2e84e11d00c.png)

<hr>

heatherdesigns commented on docroot/themes/humsci/humsci_basic/src/scss/utilities/_raised-cards.scss:
```diff
@@ -84,3 +84,88 @@
+
+// Dark Pattern and Inversion settings for hb-card, placed here because of the order of the classes when applied
+.hb-raised-cards .hb-dark-pattern .hb-card,
+.hb-dark-inversion .hb-raised-cards .views-element-container .hb-card {
```
I took a look at how the class flow in the markdown. What you have here makes sense. I can't think of any other way to inject the styles within the code above. 👍

rkelbel replied:
Thank you.

heatherdesigns commented on docroot/themes/humsci/humsci_basic/src/scss/components/_card.scss:
```diff
@@ -328,6 +348,11 @@
+
+      .hb-dark-pattern & .field-content,
+      .hb-dark-inversion .views-element-container & .field-content {
+        color: inherit;
```
This makes the month and the date inside the tile white. Perhaps remove lines 352 thru 355 and apply `color: inherit` to only `.hb-card__date-tile-day` on line 384.

rkelbel replied:
I ended up applying this just to date as suggested but used primary as it was getting that initially and I felt like being heavy-handed was needed. Hb-raised-cards have its own styling of this because of how the classes are working with each card setup and dark inversion hierarchy.

rkelbel commented:
> The weird bumpy transition of the arrows in views on hover is a result of the additional padding. I didn't investigate further into why it's happening. This is on both Colorful and Traditional.

This is a pre-existing bug, we will need to investigate this further and address it.

heatherdesigns reviewed (commented):
Excellent work! This is sooooo close!

I noticed a couple things that I didn't see before. I'm wondering if the first one is the result of the fix to the date in the event square? 🤔 

1. The hover state of the more link in views needs work when the dark color inversion toggle is applied. I did not see these issues when the utility class is applied to the view. On Traditional, when hovering on the more link the arrow remains white. It should change to match the color of the `Read More` text.
    
    ![traditional arrow hover color - inversion](https://user-images.githubusercontent.com/12678977/103666334-8273e000-4f42-11eb-8bd7-e43b13d19a68.png)

On Colorful, the text appears on hover. It should be transparent.

![colorful arrow hover - inversion](https://user-images.githubusercontent.com/12678977/103666398-97507380-4f42-11eb-9493-b15b60ffa3f3.png)

![example two on hover](https://user-images.githubusercontent.com/12678977/103666413-9c152780-4f42-11eb-9c6a-04b669c3e4b5.png)

<hr>

2. Looking at the pills under Research Areas on http://sparkbox-sandbox.suhumsci.loc/people/people-flexible-image-classes, there appears to be a border on the pill. The border is on the bottom, then appears on the top on hover. This is when the dark color inversion toggle is applied and when the utility class is applied to the view in the Colorful Theme only.

    ![pills](https://user-images.githubusercontent.com/12678977/103666719-fca46480-4f42-11eb-89b5-d1489bcde6b5.png)

<hr>

3. Add dark color inversion+pattern styles for when the `hb-display-more-link-text` utility class is added to layout builder.

![hb-display-more-link-test utility class applied](https://user-images.githubusercontent.com/12678977/103686835-65003f80-4f5d-11eb-86bb-16f87e42f5da.png)

rkelbel commented:
> 1. The hover state of the more link in views needs work when the dark color inversion toggle is applied. I did not see these issues when the utility class is applied to the view. On Traditional, when hovering on the more link the arrow remains white. It should change to match the color of the `Read More` text.

I believe I found the issue here, the arrow was changing color but it wasn't changing to tertiary highlight with the text it was using secondary highlight so it was the slightest of changes and made things look inaccurate on hover.

I have updated it to the correct color, I think this was accidentally changed when addressing other issues for colorful.
Here is a gif of me putting the tertiary highlight as the background color of the field and then using the hover state change, you can see the color transition is working and it matches the background color in its normal state there.

![alt](http://g.recordit.co/RKFOUM95yl.gif)

Additionally, on Colorful, this was a matter of hierarchy in the classes, the "not" class hover was overriding the hover for the cards more link. I have adjusted this in the mixin and it should now be working, keeping the text transparent.

![alt](http://g.recordit.co/fNdp84O3Tq.gif)

rkelbel commented:
> 1. Looking at the pills under Research Areas on http://sparkbox-sandbox.suhumsci.loc/people/people-flexible-image-classes, there appears to be a border on the pill. The border is on the bottom, then appears on the top on hover. This is when the dark color inversion toggle is applied and when the utility class is applied to the view in the Colorful Theme only.

Good catch. This was a case of the "not" class taking over all the anchors in the pills and adding the same color border as the anchors on the view.

I have made a more specific class to cover these dark inverted pills on colorful in the mixin.

![alt](http://g.recordit.co/TQMpyu3Xjv.gif)

rkelbel commented:
> 1. Add dark color inversion+pattern styles for when the `hb-display-more-link-text` utility class is added to layout builder.

This should be all set, used the same color settings for the link as the arrows on colorful when inverted.

![alt](http://g.recordit.co/zjhPXFnrOb.gif)

heatherdesigns commented on docroot/themes/humsci/humsci_basic/src/scss/components/_card.scss:
```diff
@@ -328,6 +348,11 @@
       top: 0;
       left: 0;
+
+      // .hb-dark-pattern & .field-content,
```
Let's remove this code block if we don't need it.

rkelbel replied:
Yep, Good catch!

heatherdesigns commented on docroot/themes/humsci/humsci_basic/src/scss/utilities/_display-more-link-text.scss:
```diff
@@ -13,4 +13,24 @@
   }
+
+  .hb-dark-pattern .hb-card__decorative-link a, // cards in dark mode views
+  .hb-dark-pattern .field-hs-postcard-link a { // postcards in dark mode layout
```
Can we update so that the `hb-display-more-link-text` works when the dark color inversion is selected in the theme setting as well?

rkelbel replied:
Thank you, I moved to quickly on this, it is updated now and should work with dark inversion setting.

heatherdesigns commented on docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.lists.scss:
```diff
@@ -98,6 +114,14 @@
     @include hb-ext-icon-color('black');
     background-image: none;
 
+    
```
Let's remove this extra line. It is throwing an error when you run `npm run test`.

![failing test](https://user-images.githubusercontent.com/12678977/103790444-35aa0b00-500f-11eb-9222-e01f38c259e3.png)

rkelbel replied:
Good catch, fixed up.

heatherdesigns commented on docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.lists.scss:
```diff
@@ -98,6 +114,14 @@
+    @include hb-themes(('colorful', 'airy')) {
+      .hb-dark-pattern &,
+      .hb-dark-inversion .views-element-container & {
+        @include hb-pairing-color('border-color', 'tertiary-highlight-darken-10');
```
Can we make this `border-color: transparent;` instead? On a light background the border provides a nice hover state. Since we are now on a dark background, the border is no longer necessary.

You will need to add the border: transparent for the dark inversion and dark pattern for the hover state as well to override the default border-color on line 128 below.

```
      .hb-dark-pattern &,
      .hb-dark-inversion .views-element-container & {
        border-color: transparent;
      }
```

rkelbel replied:
This is what the hover state looks like in light mode:
http://g.recordit.co/us1ONjIMWh.gif

If I add the border transparent to the focus state, the border on hover state is removed completely, which I'm fine with, I just want to clarify this is what we want.
http://g.recordit.co/UNPCJKN0n3.gif


----- 

Confirmed this via Slack conversation with Heather, just leaving that here for confirmation that we decided to go with the hover state featured here.

heatherdesigns reviewed (commented):
It's almost ready!! 🥳  I left a couple comment in the code.

heatherdesigns reviewed (commented):
Almost there!!! 🙈  Our tests are failing. Let's remove this space and then we'll be ready to roll!!

![errors](https://user-images.githubusercontent.com/12678977/104036950-0f68a480-51a2-11eb-86db-1345668da4e6.png)

Commit:
chore: compile css

heatherdesigns reviewed (approved):
APPROVED!!!! :tada: :tada: :tada:

![hallelujah](https://user-images.githubusercontent.com/12678977/104041590-49887500-51a7-11eb-989e-54fb0e5f0c31.gif)
